### PR TITLE
Implement check scope in parsing Identifier function

### PR DESCRIPTION
**Pull request by boohyunsik**
Implement checking scope feature in parseIdentifier()

### CODE REVIEW

File: parse/parser.go [hunk 1]
```diff
@@ -484,9 +484,13 @@ func parseIdentifier(buf TokenBuffer) (ast.Expression, error) {
 		return nil, ExpectError{token, Ident}
 	}
 
-	// TODO: check whether variable name exist,
-	// TODO: but do not update symbol table in this case
-	// TODO: if not, return error
+	ok := scope.Get(token.Val)
```
Comment (zeroFruit) [inline parse/parser.go]: we can inline the function

```
if scope.Get(token.Val) != nil { ...
```

File: parse/parser.go [hunk 2]
```diff
@@ -484,9 +484,13 @@ func parseIdentifier(buf TokenBuffer) (ast.Expression, error) {
...
+	if ok != nil {
+		return nil, Error{
+			token,
+			"Variable is already defined",
```
Comment (zeroFruit) [inline parse/parser.go]: We can use `DupSymError` when there is same symbol

File: parse/parser.go [hunk 3]
```diff
@@ -484,9 +484,9 @@ func parseIdentifier(buf TokenBuffer) (ast.Expression, error) {
 		return nil, ExpectError{token, Ident}
 	}
 
-	// TODO: check whether variable name exist,
-	// TODO: but do not update symbol table in this case
-	// TODO: if not, return error
+	if exist := scope.Get(token.Val); exist != nil {
```
Comment (zeroFruit) [inline parse/parser.go]: After we `token := buf.Read()` in 482 line, we should check whether this symbol exist in `scope`, if not exist in scope then should return error

Comment (zeroFruit) [reply parse/parser.go]: And plz add tcs for this logic

File: parse/parser_internal_test.go [hunk 4]
```diff
@@ -2047,6 +2118,12 @@ func TestParseExpression(t *testing.T) {
 				},
 				0,
 			},
+			func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("a", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: This does not impact on the tcs but..
the `"1"` in the `&ast.Identifier{Value: "1"}` **is not the value** of `a`, the `Value` should be name of the `a` so, should be `Value: "a"`

Comment (zeroFruit) [reply parse/parser_internal_test.go]: except this LGTM! NICE WORK!!

Comment (boohyunsik) [reply parse/parser_internal_test.go]: Ah, mistake :-)

File: parse/parser_internal_test.go [hunk 5]
```diff
@@ -1085,6 +1121,11 @@ func TestMakePrefixExpression(t *testing.T) {
 				},
 				0,
 			},
+			setupScope: func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("a", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 6]
```diff
@@ -1117,6 +1160,11 @@ func TestMakePrefixExpression(t *testing.T) {
 				},
 				0,
 			},
+			setupScope: func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("foo", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 7]
```diff
@@ -1128,6 +1176,11 @@ func TestMakePrefixExpression(t *testing.T) {
...
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 8]
```diff
@@ -1424,6 +1482,11 @@ func TestParseGroupedExpression(t *testing.T) {
 				},
 				0,
 			},
+			setupScope: func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("a", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 9]
```diff
@@ -1445,6 +1508,11 @@ func TestParseGroupedExpression(t *testing.T) {
...
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 10]
```diff
@@ -2260,6 +2375,11 @@ func TestParseIfStatement(t *testing.T) {
 				},
 				0,
 			},
+			func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("a", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 11]
```diff
@@ -2758,6 +2899,11 @@ func TestParseStatement(t *testing.T) {
 				},
 				0,
 			},
+			setupScope: func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("asdf", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 12]
```diff
@@ -2869,6 +3029,12 @@ func TestParseExpressionStatement(t *testing.T) {
 				},
 				0,
 			},
+			func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("read", &symbol.Function{Name: "read"})
+				scope.Set("x", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 13]
```diff
@@ -2885,6 +3051,13 @@ func TestParseExpressionStatement(t *testing.T) {
 				},
 				0,
 			},
+			func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("testFunction", &symbol.Function{Name: "testFunction"})
+				scope.Set("a", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
+				scope.Set("b", &symbol.Integer{Name: &ast.Identifier{Value: "2"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

File: parse/parser_internal_test.go [hunk 14]
```diff
@@ -2902,6 +3075,13 @@ func TestParseExpressionStatement(t *testing.T) {
 				},
 				0,
 			},
+			func() *symbol.Scope {
+				scope := symbol.NewScope()
+				scope.Set("testFunction", &symbol.Function{Name: "testFunction"})
+				scope.Set("a", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
+				scope.Set("b", &symbol.Integer{Name: &ast.Identifier{Value: "1"}})
```
Comment (zeroFruit) [inline parse/parser_internal_test.go]: fix here plz 👍

**Pull request merged**